app/services/authorizer.rb

```diff
@@ -1,19 +1,23 @@
 class Authorizer
+  include AuthorizerCache
+
   attr_accessor :user, :base_collection, :organization_ids, :location_ids
 
   def initialize(user, options = {})
-    @cache = HashWithIndifferentAccess.new { |h, k| h[k] = HashWithIndifferentAccess.new }
+    initialize_cache
     self.user = user
     self.base_collection = options.delete(:collection)
   end
 
-  def can?(permission, subject = nil)
+  def can?(permission, subject = nil, cache = true)
     if subject.nil?
       user.permissions.where(:name => permission).present?
     else
       return true if user.admin?
-      collection = @cache[subject.class.to_s][permission] ||= find_collection(subject.class, :permission => permission)
-      collection.include?(subject)
+      return collection_cache_lookup(subject, permission) if cache
```

Comment:
Since `user` is public accessor it can be changed for existing authorizer object, cache does not take user into account. I think to prevent security issues in future it should either invalidate the cache after user change or disable the changing of user.